### PR TITLE
Finish refactoring ParametricUMAP

### PR DESCRIPTION
**Pull request by fchollet**
This finishes the planned changes for ParametricUMAP. This PR branches off #1104. We can merge #1104 first to get a smaller diff here.

Summary:

- Supports TF, JAX, torch
- Keras 3 only
- Less code than before, and less complexity

Note that, because the changes are extensive, I'm not 100% sure whether the unit tests are quite enough to validate correctness.

### CODE REVIEW

**Comment by lmcinnes** [comment]
I'm really just waiting on a thumbs up from Tim on the other PR, and then I'll merge that and this and clean up anything left. Thanks so much for taking the time to do this.

**Comment by timsainb** [comment]
I went through everything and it looks good!

I was holding off from commenting because I wanted to think a bit about the remaining tensorflow dependency for constructing the edge dataset. We rely on tensorflow still for three things:

1) create the dataset of edges

```{python}
# create edge iterator
    edge_dataset = tf.data.Dataset.from_tensor_slices(
        (edges_to_exp, edges_from_exp)
    )
    edge_dataset = edge_dataset.repeat()
    edge_dataset = edge_dataset.shuffle(10000)
    edge_dataset = edge_dataset.batch(batch_size, drop_remainder=True)
    edge_dataset = edge_dataset.map(
        gather_X, num_parallel_calls=tf.data.experimental.AUTOTUNE
    )
    edge_dataset = edge_dataset.map(
        get_outputs, num_parallel_calls=tf.data.experimental.AUTOTUNE
    )
    edge_dataset = edge_dataset.prefetch(10)
```

2) gather the input data from the edge indices

```{python}
def gather_X(edge_to, edge_from):
        # gather data from indexes (edges) in either numpy of tf, depending on array size
        if gather_indices_in_python:
            edge_to_batch = tf.py_function(gather_index, [edge_to], [tf.float32])[0]
            edge_from_batch = tf.py_function(gather_index, [edge_from], [tf.float32])[0]
        else:
            edge_to_batch = tf.gather(X, edge_to)
            edge_from_batch = tf.gather(X, edge_from)
        return edge_to_batch, edge_from_batch
```

This is a bit lower level than I think we can abstract over with keras so far as I'm aware, so I'm wondering if @fchollet has an opinion on how to implement the dataset so that we can abstract in keras. For example, I can imagine having an input argument for which backend to use, and if the backend is tensorflow we use the current code, and if its torch, we create a torch dataset instead (and presumably for jax the same).

**Comment by fchollet** [comment]
> This is a bit lower level than I think we can abstract over with keras so far as I'm aware, so I'm wondering if @fchollet has an opinion on how to implement the dataset so that we can abstract in keras. For example, I can imagine having an input argument for which backend to use, and if the backend is tensorflow we use the current code, and if its torch, we create a torch dataset instead (and presumably for jax the same).

In Keras 3, data loading and model training/inference are completely orthogonal. You can use a `tf.data.Dataset` with a model that runs in JAX or torch. All input formats work with all backends. If you're using torch, then we create a torch DataLoader for the data independently of what was passed.

We have two options:

- We keep the current `tf.data` base pipeline. That also works with the JAX or torch backends. The only downside is that you need to have TF installed (which not required for running Keras models otherwise).
- We convert it to plain NumPy arrays (we just pass `x=edges_to_exp, y=edges_from_exp`). That also works with all backends (and it won't be less performant because prefetching is handled by `fit()`).

The second option is something that can be added independently of this PR.

**Comment by timsainb** [comment]
> The second option is something that can be added independently of this PR.
Is this possible? `edges_to_exp` and `edges_from_exp` are the indices, while the actual data that is being input to the network are the data samples. We do this since each edge is sampled up to 200 times (based on edge probability) and we wouldn't want to repeat the data samples 200 times for memory reasons. In tensorflow in the dataloader I use `tf.gather(X, edge_to)`

**Comment by fchollet** [comment]
Got it, that makes sense. So this needs to be some kind of generator, not plain NumPy arrays -- either a Python generator (not performant), a Keras `PyDataset` (backend agnostic, basically a multi-processed generator), a torch DataLoader (also a basically a multi-processed generator, but torch-specific), or a tf.data.Dataset. All of these options are valid and work with all backends.

If you want to write pure Python/Keras code that does not import any framework, you could just use a `PyDataset`. You could also do two cases that import tf or torch conditionally on the backend: `if backend() == "torch": # create a DataLoader; if backend() in ("tensorflow", "jax"): # create a tf dataset`.

**Comment by timsainb** [comment]
Ok great I can do that in a second PR @lmcinnes

**Comment by lmcinnes** [comment]
Sounds good to me.  I'll merge this then, and the other PR and if you want to handle the next step @timsainb then you are more than welcome to merge that yourself if you want.

**Comment by lmcinnes** [comment]
@fchollet I presume you would like a release to PyPI once this is all working? I have another change I would like to roll in relating to handling packed bit embeddings (as from cohere and nomic), but can ensure we get a release out pretty soon.

**Comment by fchollet** [comment]
Thanks for merging!

> @fchollet I presume you would like a release to PyPI once this is all working? I have another change I would like to roll in relating to handling packed bit embeddings (as from cohere and nomic), but can ensure we get a release out pretty soon.

Sounds good -- what's your timeline?

**Comment by lmcinnes** [comment]
I was hoping to get the binary embedding vector stuff done this weekend if possible. If I don't get that done I will just make a release as soon as Tim gets his stuff done.

**Comment by timsainb** [comment]
Hey @fchollet I have (finally) gotten around to updating everything to support torch backend. I noticed in the update you made that you switched out the way losses were compiled so that a single loss is computed in `define_loss` where previously we had each loss (e.g. global structure, reconstruction) and its weights fed directly to compile(). 

The result is now the model history doesn't log the different losses, just the overall loss. Do you have a recommendation for how to log this info in keras? e.g. using a callback or a metric

Current code: 
```{python}
def compute_loss(
    self, x=None, y=None, y_pred=None, sample_weight=None, **kwargs
):
    losses = []
    # Regularization losses.
    for loss in self.losses:
        losses.append(ops.cast(loss, dtype=keras.backend.floatx()))

    # umap loss
    losses.append(self._umap_loss(y_pred))

    # global correlation loss
    if self.global_correlation_loss_weight > 0:
        losses.append(self._global_correlation_loss(y, y_pred))

    # parametric reconstruction loss
    if self.parametric_reconstruction:
        losses.append(self._parametric_reconstruction_loss(y, y_pred))

    return ops.sum(losses)
```

previous: 
```{python}
losses = {}
loss_weights = {}

umap_loss_fn = umap_loss(
    self.batch_size,
    self.negative_sample_rate,
    self._a,
    self._b,
    self.edge_weight,
    self.parametric_embedding,
)
losses["umap"] = umap_loss_fn
loss_weights["umap"] = 1.0

if self.global_correlation_loss_weight > 0:
    losses["global_correlation"] = distance_loss_corr
    loss_weights["global_correlation"] = self.global_correlation_loss_weight
    if self.run_eagerly == False:
        # this is needed to avoid a 'NaN' error bug in tensorflow_probability (v0.12.2)
        warn("Setting tensorflow to run eagerly for global_correlation_loss.")
        self.run_eagerly = True

if self.parametric_reconstruction:
    losses["reconstruction"] = self.parametric_reconstruction_loss_fcn
    loss_weights["reconstruction"] = self.parametric_reconstruction_loss_weight

self.parametric_model.compile(
    optimizer=self.optimizer,
    loss=losses,
    loss_weights=loss_weights,
    run_eagerly=self.run_eagerly,
)
```

**Comment by timsainb** [comment]
I ended up using metrics:

```{python}
# in class UMAPModel(keras.Model).__init__()
self.global_correlation_loss_metric = keras.metrics.Mean(name="global_correlation_loss")

...

# in compute_loss()
global_correlation_loss = self._global_correlation_loss(y, y_pred)
losses.append(global_correlation_loss)
self.global_correlation_loss_metric.update_state(global_correlation_loss)
```
The only weird bit is that I'm using keras.metrics.Mean, even though the input is a single value.
